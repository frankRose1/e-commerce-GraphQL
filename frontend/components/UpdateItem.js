import React, { Component } from 'react';
import { Mutation, Query } from 'react-apollo';
import Router from 'next/router';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import Loading from './Loading';
import SubTitle from './styles/SubTitleStyles';

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      title
      description
      price
    }
  }
`;

const UPDATE_ITEM_MUTATION = gql`
  mutation UPDATE_ITEM_MUTATION(
    $id: ID!
    $title: String
    $description: String
    $price: Int
  ) {
    updateItem(
      id: $id
      title: $title
      description: $description
      price: $price
    ) {
      id
      title
      description
      price
    }
  }
`;

class UpdateItem extends Component {
  state = {};

  handleSubmit = async (e, mutation) => {
    e.preventDefault();
    const res = await mutation({
      variables: {
        id: this.props.id,
        ...this.state
      }
    });
    Router.push({
      pathname: '/item',
      query: { id: res.data.updateItem.id }
    });
  };

  handleInputChange = e => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  render() {
    return (
      <Query
        query={SINGLE_ITEM_QUERY}
        variables={{
          id: this.props.id
        }}
      >
        {({ data, loading }) => {
          if (loading) return <Loading />;
          if (!data.item)
            return <p>No item found with the ID {this.props.id}</p>;
          return (
            <Mutation mutation={UPDATE_ITEM_MUTATION} variables={this.state}>
              {(updateItem, { error, loading }) => (
                <>
                  <SubTitle>Update Item</SubTitle>
                  <Form
                    data-test='update-item'
                    method='post'
                    onSubmit={e => this.handleSubmit(e, updateItem)}
                  >
                    <Error error={error} />
                    <fieldset disabled={loading} aria-busy={loading}>
                      <label htmlFor='title'>
                        Title
                        <input
                          type='text'
                          id='title'
                          name='title'
                          placeholder='Title'
                          defaultValue={data.item.title}
                          onChange={this.handleInputChange}
                          required
                        />
                      </label>

                      <label htmlFor='price'>
                        Price
                        <input
                          type='number'
                          id='price'
                          name='price'
                          placeholder='Price'
                          defaultValue={data.item.price}
                          onChange={this.handleInputChange}
                          required
                        />
                      </label>

                      <label htmlFor='description'>
                        Description
                        <textarea
                          type='text'
                          id='description'
                          name='description'
                          placeholder='Enter a description'
                          defaultValue={data.item.description}
                          onChange={this.handleInputChange}
                          required
                        />
                      </label>

                      <button type='submit'>
                        Sav{loading ? 'ing' : 'e'} Changes
                      </button>
                    </fieldset>
                  </Form>
                </>
              )}
            </Mutation>
          );
        }}
      </Query>
    );
  }
}

export default UpdateItem;
export { UPDATE_ITEM_MUTATION, SINGLE_ITEM_QUERY };
