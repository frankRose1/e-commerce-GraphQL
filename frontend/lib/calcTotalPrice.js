/**
 * Take in a users cart and return the total cost of all items in cart
 * @param {array} cart - array of cartItems which contain data about a specific item in the DB
 * @return {integer} totalPrice - total of cart items
 */
function calcTotalPrice(cart){
  return cart.reduce((tally, cartItem) => {
    //in the case an item is deleted from the DB the cartItem will have an ID that points to a deleted item
    if (!cartItem.item) return tally;
    return tally + (cartItem.quantity * cartItem.item.price);
  }, 0);
}

export default calcTotalPrice;