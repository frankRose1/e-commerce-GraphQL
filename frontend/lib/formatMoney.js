/**
 * Format dollar amount of an item
 * @param {integer} amount - dollar amount of an item in cents
 * @return newly formatted price as a string, US currency 
 */
export default function(amount){
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }
  //for whole dollar amounts, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}