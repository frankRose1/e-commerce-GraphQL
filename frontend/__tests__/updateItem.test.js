import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import Router from 'next/router';
import wait from 'waait';
import UpdateItem, { SINGLE_ITEM_QUERY } from '../components/UpdateItem';
import { fakeItem } from '../lib/testUtils';

const updates = {
  title: 'Updated Title',
  description: 'new desc',
  price: 30000
};

const mocks = [
  {
    request: { query: SINGLE_ITEM_QUERY, variables: { id: 'abc123' } },
    result: {
      data: {
        item: { ...fakeItem() }
      }
    }
  }
];

describe('<UpdateItem />', () => {
  it('renders and matches the snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UpdateItem id='abc123' />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(
      toJSON(wrapper.find('form[data-test="update-item"]'))
    ).toMatchSnapshot();
  });

  it('displays the item information in the form inputs', async () => {
    const item = fakeItem();
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UpdateItem id='abc123' />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('input[name="title"]').props().defaultValue).toBe(
      item.title
    );
    expect(
      wrapper.find('textarea[name="description"]').props().defaultValue
    ).toBe(item.description);
    expect(wrapper.find('input[name="price"]').props().defaultValue).toBe(
      item.price
    );
  });

  it('properly updates state', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UpdateItem id='abc123' />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: { name: 'description', value: updates.description }
    });
    wrapper
      .find('input[name="title"]')
      .simulate('change', { target: { name: 'title', value: updates.title } });
    wrapper
      .find('input[name="price"]')
      .simulate('change', { target: { name: 'price', value: updates.price } });
    const component = wrapper.find('UpdateItem').instance();
    expect(component.state.title).toEqual(updates.title);
    expect(component.state.description).toEqual(updates.description);
    expect(component.state.price).toEqual(updates.price);
  });

  it('properly calls the mutation', async () => {
    //need to mock the updateItem mutation because its executed inside handleSubmit
    const updateItemMock = jest.fn().mockResolvedValue({
      data: { updateItem: { ...fakeItem() } }
    });
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UpdateItem id='abc123' />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: { name: 'description', value: updates.description }
    });
    wrapper
      .find('input[name="title"]')
      .simulate('change', { target: { name: 'title', value: updates.title } });
    wrapper
      .find('input[name="price"]')
      .simulate('change', { target: { name: 'price', value: updates.price } });
    wrapper.update();
    const component = wrapper.find('UpdateItem').instance();
    component.handleSubmit({ preventDefault: jest.fn() }, updateItemMock);
    expect(updateItemMock).toHaveBeenCalledWith({
      variables: {
        ...component.state,
        id: 'abc123'
      }
    });
  });
});
