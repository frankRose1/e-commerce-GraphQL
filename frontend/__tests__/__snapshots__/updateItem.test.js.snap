// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateItem /> renders and matches the snapshot 1`] = `
<form
  className="Form-sc-14226r0-0 elgkvZ"
  data-test="update-item"
  method="post"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="title"
    >
      Title
      <input
        defaultValue="dogs are best"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        defaultValue={5000}
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        defaultValue="dogs"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a description"
        required={true}
        type="text"
      />
    </label>
    <button
      type="submit"
    >
      Sav
      e
       Changes
    </button>
  </fieldset>
</form>
`;
